src/big/big_int.cr

```diff
@@ -403,6 +402,269 @@ struct BigInt < Int
     LibGMP.sizeinbase(self, 2).to_i
   end
 
+  # Mutating +
+  def add!(other : BigInt) : BigInt
+    if @mpz == other.@mpz
+      mul!(2)
+    else
+      LibGMP.add(mpz, self, other)
+    end
+    self
+  end
+
+  # Mutating +
+  def add!(other : Int) : BigInt
+    if other < 0
+      sub!(other.abs)
+    elsif other <= LibGMP::ULong::MAX
+      LibGMP.add_ui(mpz, self, other)
+      self
+    else
+      add!(other.to_big_i)
+    end
+  end
+
+  # Mutating -
+  def sub!(other : BigInt) : BigInt
+    if @mpz == other.@mpz
+      LibGMP.set_ui(mpz, 0)
+    else
+      LibGMP.sub(mpz, self, other)
+    end
+    self
+  end
+
+  # Mutating -
+  def sub!(other : Int) : BigInt
+    if other < 0
+      add!(other.abs)
+    elsif other <= LibGMP::ULong::MAX
+      LibGMP.sub_ui(mpz, self, other)
+      self
+    else
+      sub!(other.to_big_i)
+    end
+  end
+
+  def abs! : BigInt
+    LibGMP.abs(mpz, self)
+    self
+  end
+
+  # Mutating unary -
+  def neg! : BigInt
+    LibGMP.neg(mpz, self)
+    self
+  end
+
+  def factorial! : BigInt
+    if self < 0
+      raise ArgumentError.new("Factorial not defined for negative values")
+    elsif self > LibGMP::ULong::MAX
+      raise ArgumentError.new("Factorial not supported for numbers bigger than 2^64")
+    end
+    LibGMP.fac_ui(mpz, self)
+    self
+  end
+
+  # Mutating *
+  def mul!(other : BigInt) : BigInt
+    if @mpz == other.@mpz
+      if self > 0
+        LibGMP.pow_ui(mpz, self, 2)
+      else
+        LibGMP.pow_ui(mpz, self, 2)
+        neg!
+      end
+    else
+      LibGMP.mul(mpz, self, other)
+    end
+    self
+  end
+
+  # Mutating *
+  def mul!(other : LibGMP::IntPrimitiveSigned) : BigInt
+    LibGMP.mul_si(mpz, self, other)
+    self
+  end
+
+  # Mutating *
+  def mul!(other : LibGMP::IntPrimitiveUnsigned) : BigInt
+    LibGMP.mul_ui(mpz, self, other)
+    self
+  end
+
+  # Mutating *
+  def mul!(other : Int) : BigInt
+    mul!(other.to_big_i)
+  end
+
+  def fdiv!(other : Int::Unsigned) : BigInt
+    check_division_by_zero(other)
+    unsafe_floored_div!(other)
+  end
+
+  def fdiv!(other : Int) : BigInt
+    check_division_by_zero(other)
+
+    if other < 0
+      neg!.unsafe_floored_div!(-other)
+    else
+      unsafe_floored_div!(other)
+    end
+  end
+
+  def tdiv!(other : Int) : BigInt
+    check_division_by_zero(other)
+    unsafe_truncated_div!(other)
+  end
+
+  def unsafe_floored_div!(other : BigInt) : BigInt
+    if @mpz == other.@mpz
```

Comment:
Might be able to abstract some of this.  Something like:

```cr
def check_same(other : BigInt) : Nil
  @mpz == other.@mpz ? LibGMP.set_ui(mpz, 1) : yield
  self
end


def unsafe_floored_div!(other : BigInt) : BigInt
  self.check_same other do
    LibGMP.fdiv_q(mpz, self, other)
  end
end
```

🤷

Reply:
`LibGMP.set_ui(mpz, 1)` is only applicable in 2 out of the 7 cases though.

Reply:
Ah, nevermind then.  Not worth it.